### PR TITLE
Add _dd.agent_psr tag to PrioritySampler

### PR DESCRIPTION
**Pull request by delner**
We want to tag the priority sampled spans with the agent's rate limit. This pull request adds that tag.

### CODE REVIEW

**Comment by delner** [comment]
Something doesn't seem quite right here. It's failing tests because the `RuleSampler` is wrapped by a `PrioritySampler` and not other way around?

From the failing test:

```
Datadog::PrioritySampler.new(post_sampler: rule_sampler)
```

@marcotc Why does a priority sampler own a rule sampler by default if a rule sampler falls back to priority sampling?

File: lib/ddtrace/sampler.rb [hunk 1]
```diff
@@ -247,7 +248,12 @@ def priority_assigned_upstream?(span)
 
     def priority_sample!(span)
       preserving_sampling(span) do
-        @priority_sampler.sample!(span)
+        @priority_sampler.sample!(span).tap do
+          # We want to make sure the span is tagged with the agent-derived
+          # service rate. Retrieve this from the rate by service sampler.
+          agent_rate = @priority_sampler.sample_rate(span)
```
Comment (delner) [inline lib/ddtrace/sampler.rb]: @brettlangdon This tag is meant to be "the rate limit returned by the agent." I assume this is the rate we used to sample? If so, how is this any different than `_sample_rate` when the priority sampler is used to sample?

**Comment by marcotc** [comment]
> Why does a priority sampler own a rule sampler by default if a rule sampler falls back to priority sampling? Not sure I grasp the design here.

@delner The RuleSampler is a "simple" sampler: it decides if it wants to sample the span and then marks it as `span.sampled = true` or `span.sampled = false`.

In contrast, the PrioritySampler is a "wrapping" sampler. It knows high-level concepts specific to priority sampling, for example:
* About the upstream sampling decision and how it makes us bypass any samplers configured downstream.
* How `span.sampled` should normally be always set to `true` regardless of sampling decision.
* How `_sampling_priority_v1` is actually the place to store the sampling decision instead, when using priority sampling.

The RuleSampler doesn't know about the concept of priority sampling: it knows how to make a boolean decision and that's it. If you ask it for a decision, it will make one.
The PrioritySampler knows the "bureaucracy" required to communicate that decision to the agent.

When I had to implement these two distinct responsibilities I chose to make minimal changes to the existing PrioritySampler, as changing it introduced risk to existing users of the PrioritySampler, which is the majority of our users. Our goal was to introduce a new optional sampler without affecting existing users.

I played around the idea of duplicating the "bureaucracy" required to handle priority sampling alongside the rule sampler, but that caused too much non-trivial code and tests to be duplicated, and it felt like it would lead to diverting code for handling distributed priority sampling, which would increase maintenance burden on us.

The only idea I really liked required refactoring the existing sampler classes (PrioritySampler being one of them) to have a different contract: either a sampler is a simple sampling decision class, or they are a top-level sampling class.
Simple decision class just perform decisions. They might also attach tags/metrics to the trace to communicate how they performed their decision, but they are only required to make a decision given a trace.
Top-level sampling classes know how to deal with external factors required to properly interact with external parties that are interested in this trace. They know what information came from the outside world (an example today: distributed sampling decision), and how to properly format the new sampling decision back to the external world.

This assumes that there's more than one possible implementation of "Top-level sampling classes". If there's only one, then all sampling can be performed by "Simple decision class" and we have a very nice contract to implement. We would "promote" "Top-level sampling classes" to be a core part of the tracer in some capacity in such scenario.

TLDR: PrioritySampler the currently the class that knows how to talk to the outside world about sampling. RuleSampler (and all other sampler classes) know to make simple decisions only. Most of our users today will always want the functionality present in PrioritySampler to be active, regardless of sampling decision logic.

File: lib/ddtrace/sampler.rb (hunk 1 above)
Comment (brettlangdon) [reply lib/ddtrace/sampler.rb]: `_sample_rate` is for client sampling that tells the agent how to upscale metrics based on what is not sent to the agent.

since priority sampling always sends all traces to the agent we shouldn't set `_sample_rate`.

**Comment by delner** [comment]
@marcotc Not replicating the priority sampling code into the `RuleSampler` makes sense, but the current design causes inconsistencies with the post sampler that underlies a `PrioritySampler`. If I'm reading and understanding the implementation of priority sampling wrapping around a `RuleSampler`, that configuration actually breaks priority sampling because it doesn't implement service rates which is a core function of priority sampling.

Is it possible to create a "priority sampling rule" which is just a simple adapter for the existing priority sampler? Or maybe set the `PrioritySampler` as the default sampler for the `RuleSampler`? I might be missing something that doesn't make this plausible though...

File: lib/ddtrace/sampler.rb [hunk 2]
```diff
@@ -247,7 +248,12 @@ def priority_assigned_upstream?(span)
 
     def priority_sample!(span)
       preserving_sampling(span) do
-        @priority_sampler.sample!(span)
+        @priority_sampler.sample!(span).tap do
```
Comment (brettlangdon) [inline lib/ddtrace/sampler.rb]: There is a chance this isn't a `RateByServiceSampler`, should we add a check here?

e.g.

```
if @priority_sampler.is_a?(RateByServiceSampler) do
  span.set_metric(AGENT_RATE_METRIC_KEY, @priority_sampler.sample_rate(span))
end
```

Comment (delner) [reply lib/ddtrace/sampler.rb]: I think there were certain scenarios where this could be the case. I added the check accordingly.

**Comment by delner** [comment]
Okay, made a slight modification, and added some tests for this.

File: lib/ddtrace/sampler.rb [hunk 3]
```diff
@@ -247,7 +248,15 @@ def priority_assigned_upstream?(span)
...
+          # We want to make sure the span is tagged with the agent-derived
+          # service rate. Retrieve this from the rate by service sampler.
+          # Only do this if it was set by a RateByServiceSampler.
+          if @priority_sampler.is_a?(RateByServiceSampler)
```
Comment (brettlangdon) [inline lib/ddtrace/sampler.rb]: The problem we might have here is that if the `@priority_sampler` is a `RuleSampler` (new default https://github.com/DataDog/dd-trace-rb/blob/930d8210ab2bc6f54b35503aa41be9bb674c1c93/lib/ddtrace/tracer.rb#L409) then we still need to set this tag when the `RuleSampler` uses a `RateByServiceSampler`

Comment (marcotc) [inline lib/ddtrace/sampler.rb]: From https://github.com/DataDog/dd-trace-rb/pull/891#pullrequestreview-340722745, @brettlangdon commented:
> as long as we only set the metric when RateByServiceSampler is used, and we always set the metric regardless of whether the span is sampled by RateByServiceSampler or not

Since https://github.com/DataDog/dd-trace-rb/pull/917 was merged, the default `@priority_sampler` in this line will be a RuleSampler: https://github.com/DataDog/dd-trace-rb/blob/930d8210ab2bc6f54b35503aa41be9bb674c1c93/lib/ddtrace/tracer.rb#L409

The RuleSampler has a fallback `RateByServiceSampler` inside by default, but it only uses it if no rules match, which is also the default.

I'm thinking that one possible solution is to move this metric setting logic to `RateByServiceSampler#sample!` instead.

Comment (delner) [reply lib/ddtrace/sampler.rb]: My issue with moving it there is that not everything that samples by service rate is priority sampling... its a flawed assumption that could lead to incorrect metrics/sample rates.

Comment (brettlangdon) [reply lib/ddtrace/sampler.rb]: > not everything that samples by service rate is priority sampling

it should be, that is the only use for `RateByServiceSampler`, that is priority sampling controlled by the agent.

Comment (brettlangdon) [reply lib/ddtrace/sampler.rb]: it also doesn't necessarily matter, since the tag is meant to indicate that agent controlled sampling had a hand in deciding the sampling of a span.

Comment (delner) [reply lib/ddtrace/sampler.rb]: Would it not be possible, even if not probable, for a user to define `RateByServiceSampler` as their default sampler and set the sample rates directly for each of their services that way? Seems like a legitimate use of the component to me.

Comment (brettlangdon) [reply lib/ddtrace/sampler.rb]: I guess, it is more of... if the `RateByServiceSampler` is used to set the sampling priority (compared with setting `span.sampled`).

Which right now happens if `RateByServiceSampler` is set as the `post_sampler`/`priority_sampler` in here, *or* if the `RateByServiceSampler` is used by the `RuleSampler`

Comment (delner) [reply lib/ddtrace/sampler.rb]: I guess I'm less clear on the purpose of `_dd.agent_psr`: by definition, what is this tag supposed to represent? And how is its value used?

Comment (brettlangdon) [reply lib/ddtrace/sampler.rb]: Mostly for debugging, when you look at a raw trace in the backend you'll be able to see that the agent said to use sample rate of 0.2 to determine if this span should be kept or not.

File: lib/ddtrace/sampling/rule_sampler.rb [hunk 4]
```diff
@@ -68,7 +70,15 @@ def sample?(_span)
       end
 
       def sample!(span)
-        sampled = sample_span(span) { |s| @default_sampler.sample!(s) }
+        sampled = sample_span(span) do |s|
+          @default_sampler.sample!(s).tap do
+            # If default sampler is used and its a rate by service sampler,
+            # then add the agent rate tag for debugging.
+            if @default_sampler.is_a?(RateByServiceSampler)
+              s.set_metric(AGENT_RATE_METRIC_KEY, @default_sampler.sample_rate(span))
```
Comment (brettlangdon) [inline lib/ddtrace/sampling/rule_sampler.rb]: :+1:

File: lib/ddtrace/sampler.rb (hunk 3 above)
Comment (delner) [reply lib/ddtrace/sampler.rb]: Should be addressed now.
